### DISCUSSION TITLE
Alternatives to CTEs

### DISCUSSION THREAD

**Discussion by morgangallant** [planetscale/discussion#285] [Q&A]
Hi there — noticed that Vitess (and therefore PlanetScale) doesn't support CTEs. Are there any workarounds here that would give me roughly equivalent functionality w/o the need to use CTEs?

Here's an example:
```
create table testing (id bigint not null auto_increment, parent bigint default null, primary key(id), key idx_parent(parent));
insert into testing (parent) values (null);
insert into testing (parent) values (1);
insert into testing (parent) values (2);
insert into testing (parent) values (3);
```

Ideally, I'd be able to do a query like this (with CTEs):
```
with recursive `tree`(
  id,
  parent
) as (
  select
      id,
      parent 
    from
      testing 
    where
      id = 4
  union
  all select
      o.id,
      o.parent 
    from
      testing as o join `tree` 
        on o.id = `tree`.parent
) select
    id 
  from
    `tree`;
```

Which should ideally return:

```
4
3
2
1
```

(Or the opposite ordering, not sure here, and doesn't matter)

Is there any way of accomplishing this same functionality with PlanetScale/Vitess? Open to ideas, jank is fine!

**Comment by orware** [comment]
@morgangallant,

I don't know if you can share any additional background as to why the CTE approach may have been required for your use case before, but I think there are a few options available that shouldn't require the CTE need.

There are several books out there that discuss storing tree-like structures in the database such as [Joe Celko's Trees and Hierarchies for Smarties](https://www.elsevier.com/books/joe-celkos-trees-and-hierarchies-in-sql-for-smarties/celko/978-0-12-387733-8) as well as [Bill Karwin's SQL Antipatterns](https://pragprog.com/titles/bksap1/sql-antipatterns-volume-1/) book that is currently being revised.

In particular, Bill Karwin's book has it's Naive Trees chapter which goes over a few alternatives and this presentation of his is also a good summary:
https://www.slideshare.net/billkarwin/practical-object-oriented-models-in-sql/52-Naive_Trees_Nested_Sets_examplecommentid

Especially this slide towards the end of his presentation (slide 69):
![image](https://user-images.githubusercontent.com/931650/191366292-cbee67df-04b2-42e1-87ce-505c50cede9c.png)

The SQL Antipatterns link above contains a source code link with some SQL for the chapter on trees that may be a helpful starting point to use.

Coming from my own personal background, I observed that a lot of PHP projects never took advantage of CTEs since back in the day a lot of those features weren't available in many databases, so at least one project I'm aware of, [Joomla!](https://www.joomla.org/), switched away from using the adjacency list model similar to your example, and over to something a bit more complex like the nested sets model. Bill's Closure Table approach though I always found interesting however and may be a nice middle ground in terms of ease of use and complexity in comparison.
